### PR TITLE
Bugfix/linux task tracking

### LINKED ISSUE #10: ProcessExecutionDetector not working on PIE ELF?
Playing around with the plugins, I've created a simple test binary with the following c:

```c
#include <stdio.h>
#include <string.h>

int main() {
    char buf[32];

    fgets(buf, 32, stdin);

    if (strstr(buf, "winner") == buf) {
        printf("Win!");
    }

    return 0;
}
```

I have been writing a plugin to take in a list of addresses to avoid at runtime, instead of having to compile them separately. This works for non-PIE binaries, however the following call seems to fail on PIE ELF binaries:

Pseudo-code:
```c
m_procDetector = s2e()->getPlugin<ProcessExecutionDetector>();
m_procDetector->isTracked(state)
```

Here's what I gather for what is and isn't working:

- ELF
  - Non-PIE (working)
  - PIE
     - 64-bit (not working)
     - 32-bit (working)
- PE (unknown)
- Mach-o (unknown)

When it fails, it seems to just never return true, even though the binary appears to run.

### PR DESCRIPTION
**Pull request by adrianherrera**
Refactoring to get `LinuxMonitor::getPid` to function correctly (see https://github.com/S2E/libs2eplugins/pull/12).

The LinuxMonitor plugin currently uses the approach described at https://stackoverflow.com/questions/11961490/understanding-the-getting-of-task-struct-pointer-from-process-kernel-stack to get the current task_struct and hence the PID of the process currently running in the system. Unfortunately, this approach is only valid for kernel versions < 4.x In kernel versions > 4.x the only way to get the current task_struct is via the get_current function (arch/$(ARCH)/include/asm/current.h). This required some major refactoring to get this information to the LinuxMonitor plugin. 

We introduced a new dynamic kernel module, s2e (similar to the Windows s2e.sys driver). When loaded the s2e module sends crucial information on kernel addresses and data structures to the LinuxMonitor plugin, including the address of the current task_struct pointer. With this information, the LinuxMonitor plugin is able to correctly track the current PID in the kernel.  We can also send offset values to the LinuxMonitor plugin, which makes the plugin architecture independent and more easily ported to kernels (previously these offsets were hardcoded into the LinuxMonitor plugin itself). 

This module has the added bonus that we can correctly check if the LinuxMonitor plugin is enabled when we resume from a snapshot. Previously, the LinuxMonitor plugin was checked only at kernel boot. However, when resuming from a snapshot the LinuxMonitor plugin may be in a different state (i.e. enabled or disabled) than that when the booted was originally booted. 

Finally, we rearchitectured the S2E modifications so that they can be completely disabled via a Kconfig option.

This goes with https://github.com/S2E/guest-tools/pull/6 and fixes https://github.com/S2E/libs2eplugins/issues/10

### CODE REVIEW

File: linux-4.9.3/config-x86_64 [hunk 1]
```diff
@@ -47,6 +47,7 @@ CONFIG_DEFCONFIG_LIST="/lib/modules/$UNAME_RELEASE/.config"
 CONFIG_IRQ_WORK=y
 CONFIG_BUILDTIME_EXTABLE_SORT=y
 CONFIG_THREAD_INFO_IN_TASK=y
+CONFIG_S2E=y
```
Comment (vitalych) [inline linux-4.9.3/config-x86_64:50]: Very nice idea

File: linux-4.9.3/kernel/s2e/s2e.c [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+/// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+/// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+/// SOFTWARE.
+
+#include <linux/module.h>
+
+#include <s2e/s2e.h>
+#include <s2e/linux/linux_monitor.h>
+
+static int __init s2e_init(void) {
```
Comment (vitalych) [inline linux-4.9.3/kernel/s2e/s2e.c]: Does this conform to Linux code style? (tabs for indent, space for alignment)

File: linux-4.9.3/kernel/s2e/s2e.c [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+static void __exit s2e_exit(void) {
+    /* Nothing to do here */
+}
+
+module_init(s2e_init);
+module_exit(s2e_exit);
+
+MODULE_DESCRIPTION("S2E LinuxMonitor support");
+MODULE_LICENSE("GPL");
```
Comment (vitalych) [inline linux-4.9.3/kernel/s2e/s2e.c]: MIT

**Review by vitalych — CHANGES_REQUESTED** [review]
minor comments

**Comment by vitalych** [comment]
Is s2e module built by default?

File: linux-4.9.3/kernel/s2e/s2e.c (hunk 2 above)
Comment (adrianherrera) [reply linux-4.9.3/kernel/s2e/s2e.c]: Hmm actually no. I'll fix this.

**Comment by adrianherrera** [comment]
Yes, the module is built by default if `CONFIG_S2E` is enabled.

**Comment by adrianherrera** [comment]
Fixed s2e.c formatting. Ready to be reviewed again.

**Comment by adrianherrera** [comment]
I've also ported the kernel module to the Decree kernel.

**Review by vitalych — APPROVED** [review]

**Pull request merged**